File: src/csharp/ef-api/Controllers/AppController.cs
using EFExample.Data;
using Microsoft.AspNetCore.Mvc;

namespace EFExample.Controllers;

[ApiController]
[Route("[controller]")]
public class AppController(
  ILogger<AppController> logger,
  ResultsRepository resultsRepository
) : ControllerBase {
  [HttpGet]
  public string Get() => "Hello, World!";

  [HttpGet("/top10/{email}")]
  public async Task<List<RunnerRaceResult>> GetTop10FinishesByRunner(string email) {
    var results = await resultsRepository.Top10FinishesByRunner(email);
    return [.. results];
  }

  [HttpGet("/results/{email}")]
  public async Task<RunnerResults> GetRunnerResults(string email) {
    var result = await resultsRepository.RunnerResults(email);
    return new(
      result,
      [..result.RaceResults ?? []],
      [..result.Races ?? []]
    );
  }
}
public record RunnerResults(
  Runner Runner,
  RaceResult[] Results,
  Race[] Races
);

File: src/typescript/prisma-api/src/app.controller.ts
import { Controller, Get, Param } from '@nestjs/common';
import { AppService } from './app.service';
import { RaceResultDto as RunnerRaceResult, ResultsRepository } from './results-repository';
import { Race, RaceResult, Runner } from '@prisma/client';

@Controller()
export class AppController {
  constructor(
    private readonly appService: AppService,
    private readonly resultsRepository: ResultsRepository
  ) {}

  @Get('/top10/:email')
  async getTop10FinishesByRunner(
    @Param('email') email: string
  ): Promise<RunnerRaceResult[]> {
    return await this.resultsRepository.top10FinishesByRunner(email)
  }

  @Get('/results/:email')
  async getRunnerResult (
    @Param('email') email: string
  ) : Promise<RunnerResults> {
    const result = await this.resultsRepository.runnerResults(email)
    return {
      runner: result,
      results: result.races,
      races: result.races.flatMap(r => r.race)
    }
  }
}

type RunnerResults = {
  runner: Runner,
  results: RaceResult[],
  races: Race[]
}
